src/awkward/_connect/numba/layoutbuilder.py

```diff
@@ -0,0 +1,2547 @@
+class Empty(LayoutBuilder):
+    def __init__(self, *, parameters=None):
+        self._parameters = parameters
+
+    def __repr__(self):
+        return f"ak.numba.lb.Empty(parameters={self.parameters})"
+
+    def numbatype(self):
+        return EmptyType(numba.types.StringLiteral(self._parameters))
+
+    @property
+    def _length(self):
+        return 0
+
+    def __len__(self):
+        return self._length
+
+    @property
+    def parameters(self):
+        return self._parameters
+
+    def clear(self):
+        pass
+
+    def is_valid(self, error: str):
+        return True
+
+    def snapshot(self) -> ArrayLike:
+        return ak.Array(ak.contents.EmptyArray(parameters=self._parameters))
+
+
+class EmptyType(numba.types.Type):
+    def __init__(self, parameters):
+        super().__init__(
+            name=f"ak.numba.lb.Empty(parameters={parameters.literal_value if isinstance(parameters, numba.types.Literal) else None})"
+        )
+        self._parameters = parameters
+
+    @property
+    def parameters(self):
+        return numba.types.StringLiteral(self._parameters)
+
+    @property
+    def length(self):
+        return numba.types.int64
+
+
+@numba.extending.register_model(EmptyType)
+class EmptyModel(numba.extending.models.StructModel):
+    def __init__(self, dmm, fe_type):
+        members = [("length", fe_type.length)]
+        super().__init__(dmm, fe_type, members)
+
+
+for member in ("length",):
+    numba.extending.make_attribute_wrapper(EmptyType, member, "_" + member)
+
+
+@numba.extending.unbox(EmptyType)
+def EmptyType_unbox(typ, obj, c):
+    # fill the lowered model
+    out = numba.core.cgutils.create_struct_proxy(typ)(c.context, c.builder)
+
+    # return it or the exception
+    is_error = numba.core.cgutils.is_not_null(c.builder, c.pyapi.err_occurred())
+    return numba.extending.NativeValue(out._getvalue(), is_error=is_error)
+
+
+@numba.extending.box(EmptyType)
+def EmptyType_box(typ, val, c):
+    # get PyObject of the Empty class
+    Empty_obj = c.pyapi.unserialize(c.pyapi.serialize_object(Empty))
+
+    out = c.pyapi.call_function_objargs(Empty_obj, ())
+
+    # decref PyObjects
+    c.pyapi.decref(Empty_obj)
+
+    return out
+
+
+@numba.extending.overload_method(EmptyType, "_length_get", inline="always")
+def Empty_length(builder):
+    def getter(builder):
+        return builder._length
+
+    return getter
+
+
+@numba.extending.overload_method(EmptyType, "snapshot")
+def Empty_snapshot(builder):
+    def snapshot(builder):
+        out = np.empty(0)
+        return out
+
+    return snapshot
+
+
+########## ListOffset #########################################################
+
+
+@final
+class ListOffset(LayoutBuilder):
+    def __init__(self, dtype, content, *, parameters=None, initial=1024, resize=8.0):
+        self._offsets = GrowableBuffer(dtype=dtype, initial=initial, resize=resize)
+        self._offsets.append(0)
+        self._content = content
+        self._parameters = parameters
```

Comment:
All `dtypes` accepted from the user need to be regularized to a common form, such as NumPy's `dtype` object.

```suggestion
class ListOffset(LayoutBuilder):
    def __init__(self, dtype, content, *, parameters=None, initial=1024, resize=8.0):
        self._offsets = GrowableBuffer(dtype=numpy.dtype(dtype), initial=initial, resize=resize)
        self._offsets.append(0)
        self._content = content
        self._parameters = parameters
```